spec/index.bs

```diff
@@ -412,6 +412,36 @@ for the [=usage scope=] of this render pass.
 
 Issue(gpuweb/gpuweb#514): Document read-only states for depth views.
 
+## Synchronization ## {#programming-model-synchronization}
+
+For each [=subresource=] of a [=physical resource=], the usage is tracked on the [=Queue timeline=].
+
+Timeline could be split into a sequence of [=usage scope=]:
+within each scope the usage of the [=subresource=] stays unchanged,
+and the implementation transitions the [=subresource=] to the new usage
+at the [=usage scope=] boundaries.
+
+This specification defines the following [=usage scopes=]:
+  1. an individual copy command on a {{GPUCommandEncoder}}, such as {{GPUCommandEncoder/copyBufferToTexture()|GPUCommandEncoder.copyBufferToTexture}}.
+  2. an individual command inside a {{GPUComputePassEncoder}}, such as {{GPUProgrammablePassEncoder/setBindGroup(index, bindGroup, dynamicOffsets)|GPUProgrammablePassEncoder.setBindGroup}}.
```

Comment:
Actually isn't it just dispatch and dispatch indirect that are usage scopes? The rest are just setting state for these two commands.

Reply:
This is a great question to raise. The reason I went for any command here is apposed to just the draws, is because it would be awkward to allow bad state to be bound even if it's not used.

So, first, I think we are going to be specifying the exact resources used by a dispatch as those that are covered by the bind groups attached to slots expected by the layout of the active pipeline. So there might be bindings *not* included/used by a dispatch, and similarly for the draws.

What happens if, for example, the user is recording a render pass, and they bind a group that uses one of the current render targets as STORAGE? It seems to me that making this an error versus non error based on the dependencies of a draw() would make up for a weaker API. It would be nice to specify that, in particular, everything you bind within a render pass becomes part of the usage scope for this pass.

Reply:
Thanks for the detailed explanation. It makes a lot of sense for the render pass. For the dispatch we'd want to consider the resources used by that dispatch and only this one so the wording might get a bit more complicated.

Reply:
What happens if the user does `setBindGroup()` with one that is self-conflicting? I.e. it has different mutable bindings for the same resource? If we spec the usage scopes for `dispatch` only, we'd be ignoring these bindings, potentially.

Reply:
I'm not sure what's best here. Both validating the `setBindGroup` by itself and not doing it seem weird.

Reply:
Would be nice to be somewhat consistent between the compute and render passes wrt the usage scopes, even if the scopes are different. I.e. it would be good to say that the usage scope of a render pass is the union of the usage scopes of its commands, and for the compute pass it's just not a union.

So the same kind of problem can be seen in render passes: if the user calls `setVertexData` we have a choice (in how we define the usage scopes):
  1. Every command by itself is a usage scope, and the render pass considers the union of all usage scopes of its commands. This is what the current PR has (or at least aims to). The consequence here is that we consider the vertex buffer to be used by the pass (regardless of what the draw calls would do).
  2. Only draws, dispatches, and copies are considered to be the usage scopes (and a render pass is a union of its draw usage scopes). This means that on each `draw()` we would have to traverse the dependencies and consider the usage scopes. In this case, it will be up to the *active* render pipeline layout at the next `draw()` to decide whether the vertex buffer is considered to be used.

I have a slight preference towards (1) for making the validation simpler, which may allow us to have higher draw call throughput in the end.

Reply:
I think it makes sense that calling setVertexBuffer will mark the buffer as used for that pass even if it's not actually accessed by the draw call. It's up to the developer to prevent that from happening if they want to use a vertex buffer as something else during that pass.